cilium/cilium
# Add option to redact http headers

This PR adds the `http-headers` part of https://github.com/cilium/cilium/issues/23887.

This is to provide the option to redact the values of http headers in http flows.

The specific feature can be configured by providing the following values for example:

```yml
hubble:
  redact:
    enabled: true
    http:
      headers:
        allow:
          - traceparent
          - tracestate
          - Cache-Control
        deny: []
```

Note_1: `allow` list means that these headers will not be redacted
Note_2: `deny` list means that these headers will be redacted
Note_3: only one of these 2 lists can be defined. If both defined then `allow` list has priority and `deny` will be ignored.

The filtering logic/algorithm is described at: https://github.com/cilium/cilium/pull/26724/commits/4c3d701f92a4dbeb533b71d404bf0fbaadce352f#diff-1df644554f9d2598342e7c4d8f9f3ae5af5fd29b52ed0dc88e01ea986f218a36R88-R98

I was not 100% sure what should be the default behavior when just `hubble.redact` is enabled but none of the lists are defined so I had it to redact everything by default. That's just my proposal but let's see what we can get from the review here.

### How to test this manually

1. In the root directory of the project edit https://github.com/cilium/cilium/blob/23bac97b3aa7a34e9d03f7d901bd77a124378446/contrib/testing/kind-values.yaml to add the following lines:
```yaml
hubble:
  redact:
    enabled: true
    http:
      headers:
        allow:
          - traceparent
          - tracestate
          - Cache-Control
```
2. Deploy cilium from source with `make kind-image && make kind-install-cilium`
3. Enable hubble with `cilium hubble enable --chart-directory=/home/chrismark/go/src/github.com/cilium/cilium/install/kubernetes/cilium`
4. Apply a L7 example endpoint with `kubectl apply -f https://raw.githubusercontent.com/cilium/cilium/1.13.3/examples/minikube/sw_l3_l4_l7_policy.yaml`
5. Port forward hubble relay svc with `kubectl port-forward -n kube-system svc/hubble-relay --address 0.0.0.0 --address :: 4245:80`
6. Deploy the sample app with `kubectl create -f https://raw.githubusercontent.com/cilium/cilium/1.13.3/examples/minikube/http-sw-app.yaml`. From https://docs.cilium.io/en/stable/gettingstarted/demo/#starwars-demo
7. Observe the flows with `hubble observe -o jsonpb --pod deathstar --protocol http -f | jq`
8. Execute a curl request with `kubectl exec tiefighter -- curl -H "Content-Type2: application/json" -s -XPOST -v "http://deathstar.default.svc.cluster.local/v1/request-landing" -H "X-Powered-By: ReqBin HTTP Client" -H "tracestate: foobar"`
9. Verify that headers' values are replaced with `HUBBLE_REDACTED` except of the `Accept` one like in the following:
```json
{
  "flow": {
    "time": "2023-09-02T08:14:26.498165607Z",
    "uuid": "5b70099c-e482-4c43-bb74-93e05933c21d",
    "verdict": "FORWARDED",
    "IP": {
      "source": "10.244.1.231",
      "destination": "10.244.1.227",
      "ipVersion": "IPv4"
    },
    "l4": {
      "TCP": {
        "source_port": 53704,
        "destination_port": 80
      }
    },
    "source": {
      "ID": 3464,
      "identity": 27221,
      "namespace": "default",
      "labels": [
        "k8s:app.kubernetes.io/name=tiefighter",
        "k8s:class=tiefighter",
        "k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default",
        "k8s:io.cilium.k8s.policy.cluster=kind-kind",
        "k8s:io.cilium.k8s.policy.serviceaccount=default",
        "k8s:io.kubernetes.pod.namespace=default",
        "k8s:org=empire"
      ],
      "pod_name": "tiefighter"
    },
    "destination": {
      "ID": 3,
      "identity": 56737,
      "namespace": "default",
      "labels": [
        "k8s:app.kubernetes.io/name=deathstar",
        "k8s:class=deathstar",
        "k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default",
        "k8s:io.cilium.k8s.policy.cluster=kind-kind",
        "k8s:io.cilium.k8s.policy.serviceaccount=default",
        "k8s:io.kubernetes.pod.namespace=default",
        "k8s:org=empire"
      ],
      "pod_name": "deathstar-7848d6c4d5-7dwtj",
      "workloads": [
        {
          "name": "deathstar",
          "kind": "Deployment"
        }
      ]
    },
    "Type": "L7",
    "node_name": "kind-kind/kind-worker",
    "l7": {
      "type": "REQUEST",
      "http": {
        "method": "POST",
        "url": "http://deathstar.default.svc.cluster.local/v1/request-landing",
        "protocol": "HTTP/1.1",
        "headers": [
          {
            "key": ":scheme",
            "value": "HUBBLE_REDACTED"
          },
          {
            "key": "Accept",
            "value": "HUBBLE_REDACTED"
          },
          {
            "key": "Content-Type2",
            "value": "HUBBLE_REDACTED"
          },
          {
            "key": "Tracestate",
            "value": "foobar"
          },
          {
            "key": "User-Agent",
            "value": "HUBBLE_REDACTED"
          },
          {
            "key": "X-Envoy-Expected-Rq-Timeout-Ms",
            "value": "HUBBLE_REDACTED"
          },
          {
            "key": "X-Envoy-Internal",
            "value": "HUBBLE_REDACTED"
          },
          {
            "key": "X-Powered-By",
            "value": "HUBBLE_REDACTED"
          },
          {
            "key": "X-Request-Id",
            "value": "HUBBLE_REDACTED"
          }
        ]
      }
    },
    "event_type": {
      "type": 129
    },
    "traffic_direction": "INGRESS",
    "is_reply": false,
    "Summary": "HTTP/1.1 POST http://deathstar.default.svc.cluster.local/v1/request-landing"
  },
  "node_name": "kind-kind/kind-worker",
  "time": "2023-09-02T08:14:26.498165607Z"
}
```
